### PR TITLE
U4-9898 Reset password should unlock a locked account

### PR DESCRIPTION
**Pull request by nul800sebastiaan**
Fixes [the issue](http://issues.umbraco.org/issue/U4-9898).

It also allows password reset mails to be sent when a user is locked out, this was previously blocked.

### CODE REVIEW

**Comment by dampee** [comment]
If you use await you really should use .ConfigureAwait(false) on framework code to avoid odd problems.
Otherwise exceptions could be swallowed and methods return preemptive...

Op 20 mei 2017 2:37 p.m. schreef Sebastiaan Janssen <notifications@github.com>:

Don't merge yet!

There's a problem here.. the Lockout property in the database (UserNoConsole) is not being reset and I don't understand why. The UpdateAsync succeeds and the AccessFailedCount property DOES indeed get saved.

Other than that, it fixes the issue<http://issues.umbraco.org/issue/U4-9898>.

It also allows password reset mails to be sent when a user is locked out, this was previously blocked.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/umbraco/Umbraco-CMS/pull/1956

Commit Summary

  *   U4-9898 Reset password should unlock a locked account

File Changes

  *   M src/Umbraco.Web/Editors/AuthenticationController.cs<https://github.com/umbraco/Umbraco-CMS/pull/1956/files#diff-0> (22)

Patch Links:

  *   https://github.com/umbraco/Umbraco-CMS/pull/1956.patch
  *   https://github.com/umbraco/Umbraco-CMS/pull/1956.diff

File: src/Umbraco.Web/Editors/AuthenticationController.cs [hunk 1]
```diff
@@ -303,9 +303,27 @@ public UserDetail GetCurrentUser()
             var result = await UserManager.ResetPasswordAsync(model.UserId, model.ResetCode, model.Password);
             if (result.Succeeded)
             {
+                var lockedOut = await UserManager.IsLockedOutAsync(model.UserId);
```
Comment (Shazwazza) [inline src/Umbraco.Web/Editors/AuthenticationController.cs]: The property `LockoutEnabled` is incorrect. This property is for something that we don't support. In ASP.NET identity you can specify whether locking out a user is allowed on a per user basis - we do not support this so changing this flag does nothing.

Further to that you shouldn't manipulate those properties directly, the user manager has methods to do this, for example if you did want to modify this you'd use `UserManager.SetLockoutEnabled(true/false)` but in Umbraco this serves no purpose, see: https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Core/Security/BackOfficeUserStore.cs#L611 

A user is determined to be locked out purely by the date/time assigned to the lockout date: https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Core/Security/BackOfficeUserStore.cs#L542 

which you can set with the `UserManager.SetLockoutEndDateAsync` or `UserManager.SetLockoutEndDate`

**Comment by nul800sebastiaan** [comment]
Updated.

@dampee As far as I understand it's only relevant to set `ConfigureAwait(false)` on background tasks. 

@Shazwazza Why don't we implement `UserManager.SetLockoutEnabled` and set the locked out date from that? it's pretty misleading. How are mere mortals supposed to figure this out?
Is it OK to set the LockoudEnd date to `DateTime.Now`? It works for me, and can't think of any timezone issues.

I'm fine with not manipulating those properties directly but now I guess it will be 2 database hits instead of one, second one for `ResetAccessFailedCount`.

**Comment by Shazwazza** [comment]
@dampee + @nul800sebastiaan ConfigureAwait(false) should be used whenever developing code libraries for re-use in other people's applications. Reason being is because if you don't specify false you may capture a synchronization context depending on when your code is being used (i.e. request or background thread). You don't  need specify this in things like controllers since controller code is not code library code that is re-used, it's application code.

@nul800sebastiaan I didn't write the ASP.NET identity framework so I have no control over what SetLockoutEnabled is intended for- this is part of ASP.NET identity *not* Umbraco and is for a specific reason as I stated -> to be able to turn on/off lockout functionality per user which we don't support, it's *not* to turn on/off the actual lockout per user. see https://msdn.microsoft.com/en-us/library/dn613149%28v=vs.108%29.aspx?f=255&MSPPError=-2147217396 

> "Sets whether the user can be locked out"

**Comment by Shazwazza** [comment]
@nul800sebastiaan you should set the lockout end date to DateTimeOffset.Now not DateTime.Now since it's expecting DateTimeOffset

**Comment by nul800sebastiaan** [comment]
Ah okay, I didn't realize. It's only for determining whether someone can or cannot be locked out. Confusing. Okay, all good there!

**Comment by nul800sebastiaan** [comment]
Sorry, yes, I meant `DateTimeOffset.Now`, that's what I did :)

**Comment by Shazwazza** [comment]
Yup that should be perfect then

**Pull request merged**